parcels/_interpolation.py

```diff
@@ -0,0 +1,265 @@
+    nb_land = np.sum(land)
+    if nb_land == 8:
+        return 0
+    elif nb_land > 0:
+        val = 0
+        w_sum = 0
+        for k in range(2):
+            for j in range(2):
+                for i in range(2):
+                    distance = pow((ctx.zeta - k), 2) + pow((ctx.eta - j), 2) + pow((ctx.xsi - i), 2)
+                    if np.isclose(distance, 0):
+                        if land[k][j][i] == 1:  # index search led us directly onto land
+                            return 0
+                        else:
+                            return ctx.data[ctx.ti, ctx.zi + k, ctx.yi + j, ctx.xi + i]
+                    elif land[k][j][i] == 0:
+                        val += ctx.data[ctx.ti, ctx.zi + k, ctx.yi + j, ctx.xi + i] / distance
+                        w_sum += 1 / distance
+        return val / w_sum
+    else:
+        data = ctx.data[ctx.ti, ctx.zi, :, :]
+        f0 = (
+            (1 - ctx.xsi) * (1 - ctx.eta) * data[ctx.yi, ctx.xi]
+            + ctx.xsi * (1 - ctx.eta) * data[ctx.yi, ctx.xi + 1]
+            + ctx.xsi * ctx.eta * data[ctx.yi + 1, ctx.xi + 1]
+            + (1 - ctx.xsi) * ctx.eta * data[ctx.yi + 1, ctx.xi]
+        )
+        data = ctx.data[ctx.ti, ctx.zi + 1, :, :]
+        f1 = (
+            (1 - ctx.xsi) * (1 - ctx.eta) * data[ctx.yi, ctx.xi]
+            + ctx.xsi * (1 - ctx.eta) * data[ctx.yi, ctx.xi + 1]
+            + ctx.xsi * ctx.eta * data[ctx.yi + 1, ctx.xi + 1]
+            + (1 - ctx.xsi) * ctx.eta * data[ctx.yi + 1, ctx.xi]
+        )
+        return (1 - ctx.zeta) * f0 + ctx.zeta * f1
+
+
+"""
+elif self.interp_method in ["linear", "bgrid_velocity", "bgrid_w_velocity", "partialslip", "freeslip"]:
+    if self.interp_method == "bgrid_velocity":
+        if self.gridindexingtype == "mom5":
+            zeta = 1.0
+        else:
+            zeta = 0.0
+    elif self.interp_method == "bgrid_w_velocity":
+        eta = 1.0
+        xsi = 1.0
+    data = self.data[ti, zi, :, :]
+    f0 = (
+        (1 - xsi) * (1 - eta) * data[yi, xi]
+        + xsi * (1 - eta) * data[yi, xi + 1]
+        + xsi * eta * data[yi + 1, xi + 1]
+        + (1 - xsi) * eta * data[yi + 1, xi]
+    )
+    if self.gridindexingtype == "pop" and zi >= self.grid.zdim - 2:
+        # Since POP is indexed at cell top, allow linear interpolation of W to zero in lowest cell
+        return (1 - zeta) * f0
+    data = self.data[ti, zi + 1, :, :]
+    f1 = (
+        (1 - xsi) * (1 - eta) * data[yi, xi]
+        + xsi * (1 - eta) * data[yi, xi + 1]
+        + xsi * eta * data[yi + 1, xi + 1]
+        + (1 - xsi) * eta * data[yi + 1, xi]
+    )
+    if self.interp_method == "bgrid_w_velocity" and self.gridindexingtype == "mom5" and zi == -1:
+        # Since MOM5 is indexed at cell bottom, allow linear interpolation of W to zero in uppermost cell
+        return zeta * f1
+    else:
+        return (1 - zeta) * f0 + zeta * f1
+"""
```

Comment:
This refactoring isn't as simple as "moving the code" like the other ones, as this needs to be separated into a couple interpolation functions. Do you want me to be rigorous with the refactoring here (i.e., create test cases for this isolated code, do the refactor and make sure they still pass), or should I do the refactoring relying on the existing test suite?